### PR TITLE
35012 limit build io

### LINKED ISSUE #35012: docker build limit io disk
Would there be any way to limit the disk IO for `docker build` ?
Something like we have in `docker run`

```
--device-read-bps Limit read rate from a device 
--device-write-bps Limit write rate to a device
--device-read-iops Limit read rate (IO per second) to a device  
--device-write-iops	Limit write rate (IO per second) to a device 
```

### PR DESCRIPTION
**Pull request by bealerj2260**
Closes #35012 

**- What we did**
Added the six options to limit read/write speeds to a device that are present in docker run to docker build. There will need to be cli changes to use this feature. We have already completed that work and will open a PR against the cli repo once this gets merged (or earlier if maintainers would prefer that). 

**- How I did it**
Added the additional parameters to the build options, and host config. 

**- How to verify it**
This has been difficult to verify as limiting the block devices is not easily observable for some reason, but you can see that the limits are applied to the build container by checking the cgroup in your docker system files `/sys/fs/cgroup/blkio/docker/<container_long_id>/blkio.throttle.<command-name>`.

**- Description for the changelog**
Add device throttle options to docker build

### CODE REVIEW

File: api/types/client.go [hunk 1]
```diff
@@ -141,26 +141,32 @@ func (h *HijackedResponse) CloseWrite() error {
...
+	Isolation            container.Isolation
+	CPUSetCPUs           string
+	CPUSetMems           string
+	CPUShares            int64
+	CPUQuota             int64
+	CPUPeriod            int64
+	BlkioReadBpsDevice   string
+	BlkioWriteBpsDevice  string
+	BlkioReadIOpsDevice  string
+	BlkioWriteIOpsDevice string
```
Comment (kolyshkin) [inline api/types/client.go]: Perhaps drop the `Device` suffix, as it looks redundant and doesn't add any more clarity.

File: builder/dockerfile/internals.go [hunk 2]
```diff
@@ -546,6 +554,62 @@ func hostConfigFromOptions(options *types.ImageBuildOptions) *container.HostConf
...
 }
 
+func throttleDeviceFromOptionWithSize(option string) []*blkiodev.ThrottleDevice {
+	configuration := strings.Split(option, ":")
+	if len(configuration) != 2 {
+		return nil
+	}
+	var rate int
+	var err error
+	if strings.HasSuffix(configuration[1], "kb") {
```
Comment (kolyshkin) [inline builder/dockerfile/internals.go]: This is basically repeating the functionality of [docker/go-units](https://github.com/docker/go-units), which is used by other similar places. Please use it here, too.

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Comment by bealerj2260** [comment]
@kolyshkin Sorry for the delay! I've renamed those fields and made use of the library that you mentioned.

**Comment by olljanat** [comment]
Derek add label: status/failing-ci

**Comment by thaJeztah** [comment]
Looks like this went stale, and we're not sure if this is something to add or too much bells and whistles, so I'll close this one for now. Feel free to continue the conversation if there's strong reasons for adding this functionality

**Pull request closed without merging**